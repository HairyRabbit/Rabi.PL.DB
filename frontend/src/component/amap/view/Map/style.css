/* -*- mode: css -*- */
/* -*- coding: utf-8 -*- */

:global(.amap-logo) {
  display: none !important;
}

:global(.amap-copyright) {
  display: none !important;
}
