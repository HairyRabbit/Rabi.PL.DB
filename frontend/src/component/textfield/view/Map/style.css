/* -*- mode: css -*- */
/* -*- coding: utf-8 -*- */

@value width: 200px;
@value height: 200px;

.part {
  fill: #ddd;
}

.part:hover,
.active {
  fill: #803C80;
}

.mesh {
  fill: none;
  stroke: #fff;
  stroke-linejoin: round;
  stroke-linecap: round;
}



