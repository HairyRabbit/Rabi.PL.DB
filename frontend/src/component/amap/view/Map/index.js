// -*- mode: js-jsx -*-
// -*- coding: utf-8 -*-
// @flow

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import loadMap from '../../lib/load-map'
import mapPositionToLngLat from '../../lib/map-position-to-lnglat'
import bindEvents from '../../lib/bind-events'
import style from './style.css'

function normalizeMapEvent(event: MapEvent): string {
  switch (event) {
    case 'onDoubleClick':
      return 'onDblClick'
    case 'onContextMenu':
      return 'onRightClick'
    case 'onDrag':
      return 'onDragging'
    default:
      return event
  }
}

function mapToMapEvent(event: MapEvent): string {
  return normalizeMapEvent(event).slice(2).toLowerCase()
}

class Map extends Component {
  constructor(props: Prop) {
    super(props)
    this.state = { mount: false }
    this.events = []
    this.AMap = null
    this.map = null
  }
  checkInfoWindow() {
    // TODO
  }
  componentWillReceiveProps(nextProps) {
    if (nextProps.zoom !== this.props.zoom) {
      this.map.setZoom(nextProps.zoom)
    }

    if (nextProps.center !== this.props.center) {
      if (typeof nextProps.center !== 'string') {
        this.map.setCenter(nextProps.center)
      } else {
        mapLocationToPosition(nextProps.center).then(location => {
          if (location) {
            this.map.setCenter(location.position)
          }
        })
      }
    }
  }

  load() {
    const { center, zoom } = this.props

    return new Promise((resolve, reject) => {
      mapLocationToPosition(center)
        .then(location => {
          this.map = new AMap.Map(this.container, {
            center: location && location.position,
            zoom,
            mapStyle: 'amap://styles/whitesmoke'
          })
        })
        .then(resolve)
        .catch(reject)
    })
  }
  componentDidMount() {
    const ready = AMap => {
      this.AMap = AMap
      return new Promise((resolve, reject) => {
        this.load()
          .then(() => {
            this.events = bindEvents(AMap, this.map, this.props)
            this.setState({ mount: true })
          })
          .then(resolve)
          .catch(reject)
      })
    }
    if (!process.env.AMAP_KEY) {
      throw new Error(`Not Found AMAP_KEY defined.`)
    }
    loadMap(process.env.AMAP_KEY, ['AMap.Geocoder']).then(ready).catch(err => {
      throw err
    })
  }
  componentWillUnmount() {
    Object.keys(events).forEach(key => {
      AMap.event.removeListener(events[key])
      delete events[key]
    })
    this.map.clearMap()
    this.map.clearInfoWindow()
    this.map.removeControl()
    this.map.destroy()
  }
  getChildContext() {
    return {
      AMap: this.AMap,
      map: this.map
    }
  }
  render() {
    const { width, height, children } = this.props

    return (
      <div
        ref={ref => (this.container = ref)}
        style={{ width: width + 'px', height: height + 'px' }}
      >
        {this.state.mount ? children : null}
      </div>
    )
  }
}

Map.childContextTypes = {
  AMap: PropTypes.object,
  map: PropTypes.object
}

function mapLocationToPosition(center) {
  return new Promise(function(resolve, reject) {
    if (!AMap && !AMap.Geocoder) {
      reject(new Error('Not Found AMap or AMap.Geocoder'))
    } else if (typeof center === 'string') {
      const geo = new AMap.Geocoder()
      geo.getLocation(center, function(status, response) {
        const { info, geocodes } = response
        if (status !== 'complete' || info !== 'OK' || geocodes.length === 0) {
          return resolve(undefined)
        } else {
          const [result] = geocodes
          const { location } = result
          resolve({
            position: location
          })
        }
      })
    } else {
      resolve(center)
    }
  })
}

export default Map
