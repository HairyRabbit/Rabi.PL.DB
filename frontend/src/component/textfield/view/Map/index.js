// -*- mode: js-jsx -*-
// -*- coding: utf-8 -*-
// @flow

import React, { Component } from 'react'
import { geoPath, geoMercator } from 'd3-geo'
import * as topojson from 'topojson'
import style from './style.css'

type Prop = {}

function Map(props) {
  const { onSelect, width, height, children } = props
  // const json = require('map/province/shanxi1.json')
  const activedItem = children.find(n => n.active)
  const json = require('map/china.json')
  const objects = topojson.feature(json, json.objects.foo)
  const path = geoPath().projection(
    geoMercator().fitSize([width, height], objects)
  )
  const meshPath = topojson.mesh(json, json.objects.foo, function(a, b) {
    return a !== b
  })
  const mapPaths = objects.features.map((d, idx) => {
    const isActive = activedItem.value === d.properties.name
    const activedStyle = isActive
      ? [style.part, style.active].join(' ')
      : style.part
    return (
      <path
        key={idx}
        onMouseDown={() => onSelect(d.properties.name)}
        className={activedStyle}
        d={path(d)}
      />
    )
  })

  return (
    <svg width={width} height={height}>
      <g>{mapPaths}</g>
      <path className={style.mesh} d={path(meshPath)} />
    </svg>
  )
}

export default Map
