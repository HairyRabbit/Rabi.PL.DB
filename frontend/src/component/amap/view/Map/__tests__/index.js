// -*- mode: js-jsx -*-
// -*- coding: utf-8 -*-

import React from 'react'
import { shadow } from 'enzyme'
import renderer from 'react-test-renderer'
import Map from '../'

describe('<Map /> snapshots', () => {
  const component = renderer.create(<Map />)

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

describe('<Map /> props', () => {})
