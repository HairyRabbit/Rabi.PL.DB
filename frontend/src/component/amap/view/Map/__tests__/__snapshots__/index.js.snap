// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`undefined 1`] = `
<div
  className="container"
  style={Object {}}
/>
`;
